syntax = "proto3";

package pymongoRPC;

service MongoDBService {
  rpc Find(FindRequest) returns (FindResponse);
  rpc FindOne(FindOneRequest) returns (FindOneResponse);
  rpc InsertOne(InsertOneRequest) returns (InsertOneResponse);
  rpc InsertMany(InsertManyRequest) returns (InsertManyResponse);
  rpc UpdateOne(UpdateOneRequest) returns (UpdateOneResponse);
  rpc DeleteMany(DeleteManyRequest) returns (DeleteManyResponse);
  rpc CreateIndex(CreateIndexRequest) returns (CreateIndexResponse);
}

message FindRequest {
  string collection = 1;
  string filter = 2; // JSON
}

message FindResponse {
  string documents = 1; // JSON
}

message FindOneRequest {
  string collection = 1;
  string filter = 2; // JSON
}

message FindOneResponse {
  string document = 1; // JSON
}

message InsertOneRequest {
  string collection = 1;
  string document = 2; // JSON
}

message InsertOneResponse {
  string insertedId = 1;
}

message InsertManyRequest {
  string collection = 1;
  string documents = 2; //JSON
}

message InsertManyResponse {
  string insertedIds = 1;
}

message UpdateOneRequest {
  string collection = 1;
  string filter = 2; // JSON
  string update = 3; // JSON
  bool upsert = 4;
}

message UpdateOneResponse {
  int32 matchedCount = 1;
  int32 modifiedCount = 2;
}

message DeleteManyRequest {
  string collection = 1;
  string filter = 2; // JSON
}

message DeleteManyResponse {
  int32 deletedCount = 1;
}

message CreateIndexRequest {
  string collection = 1;
  string keys = 2; // JSON
  string name = 3;
  string default_language = 4;
}

message CreateIndexResponse {
  string indexName = 1;
}
